src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DataAnnotationsModelMetadataDetailsProvider.cs

```diff
@@ -0,0 +1,167 @@
+            // ApplyFormatInEditMode==true.
+            if (displayFormatAttribute == null && dataTypeAttribute != null)
+            {
+                displayFormatAttribute = dataTypeAttribute.DisplayFormat;
+            }
+
+            // ConvertEmptyStringToNull
+            if (displayFormatAttribute != null)
+            {
+                details.ConvertEmptyStringToNull = displayFormatAttribute.ConvertEmptyStringToNull;
+            }
+
+            // DataTypeName
+            if (dataTypeAttribute != null)
+            {
+                details.DataTypeName = dataTypeAttribute.GetDataTypeName();
+            }
+            else if (displayFormatAttribute != null && !displayFormatAttribute.HtmlEncode)
+            {
+                details.DataTypeName = DataType.Html.ToString();
+            }
+
+            // Description
+            if (displayAttribute != null)
+            {
+                details.Description = displayAttribute.Description;
+            }
+
+            // DisplayFormat
+            if (displayFormatAttribute != null)
+            {
+                details.DisplayFormatString = displayFormatAttribute.DataFormatString;
+            }
+
+            // DisplayName
+            if (displayAttribute != null)
+            {
+                details.DisplayName = displayAttribute.Name;
+            }
+
+            if (displayFormatAttribute != null && displayFormatAttribute.ApplyFormatInEditMode)
+            {
+                details.EditFormatString = displayFormatAttribute.DataFormatString;
+            }
+
+            // HasNonDefaultEditFormat
+            if (!string.IsNullOrEmpty(displayFormatAttribute?.DataFormatString) &&
+                displayFormatAttribute?.ApplyFormatInEditMode == true)
+            {
+                // Have a non-empty EditFormatString based on [DisplayFormat] from our cache.
+                if (dataTypeAttribute == null)
+                {
+                    // Attributes include no [DataType]; [DisplayFormat] was applied directly.
+                    details.HasNonDefaultEditFormat = true;
+                }
+                else if (dataTypeAttribute.DisplayFormat != displayFormatAttribute)
+                {
+                    // Attributes include separate [DataType] and [DisplayFormat]; [DisplayFormat] provided override.
+                    details.HasNonDefaultEditFormat = true;
+                }
+                else if (dataTypeAttribute.GetType() != typeof(DataTypeAttribute))
+                {
+                    // Attributes include [DisplayFormat] copied from [DataType] and [DataType] was of a subclass.
+                    // Assume the [DataType] constructor used the protected DisplayFormat setter to override its
+                    // default.  That is derived [DataType] provided override.
+                    details.HasNonDefaultEditFormat = true;
+                }
+            }
```

Comment:
nit: since you're now setting everything up at once, might be simpler to merge this with the `displayFormatAttribute = dataTypeAttribute.DisplayFormat` special case